core/src/test/scala/integration/kafka/api/ProducerFailureHandlingTest.scala

```diff
@@ -175,23 +176,29 @@ class ProducerFailureHandlingTest extends KafkaServerTestHarness {
   }
 
   /**
-   *  The send call with invalid partition id should throw KafkaException caused by IllegalArgumentException
-   */
+    * Send with invalid partition id should throw KafkaException when partition is higher than the
+    * upper bound of partitions and IllegalArgumentException when partition is negative
+    */
   @Test
   def testInvalidPartition() {
-    // create topic
+    // create topic with partition range [0...0]
     TestUtils.createTopic(zkUtils, topic1, 1, numServers, servers)
 
-    // create a record with incorrect partition id, send should fail
-    val record = new ProducerRecord[Array[Byte],Array[Byte]](topic1, new Integer(1), "key".getBytes, "value".getBytes)
-    intercept[IllegalArgumentException] {
-      producer1.send(record)
+    // create a record with incorrect partition id (higher than the number of partitions), send should fail
+    val higherRecord = new ProducerRecord[Array[Byte],Array[Byte]](topic1, new Integer(1), "key".getBytes, "value".getBytes)
+    intercept[KafkaException] {
+      producer1.send(higherRecord)
     }
-    intercept[IllegalArgumentException] {
-      producer2.send(record)
+    intercept[KafkaException] {
+      producer2.send(higherRecord)
```

Comment:
Same here, but with unit tests I err on the side of adding more rather than removing some unless it's very important.

At a closer look, producers differ at the number of acks they require. I would leave all three. Doesn't cost.